### PR TITLE
fix: unserialize makes copyRecords without problematic assign semantics

### LINKED ISSUE #1303: Unserialization creates copyRecord with assignment rather than define semantics
This fails in two ways:
   * If a copyRecord contains a property that would override a non-enabled frozen data property inherited from `Object.prototype`.
   * If a copyRecord contains a property named `__proto__` is a system in which the `Object.prototype.__proto__` accessor property still exists. See https://github.com/Agoric/agoric-sdk/pull/6349

### PR DESCRIPTION
**Pull request by erights**
fixes #1303 

Includes new test cases that I did see fail before the code was fixed.

### CODE REVIEW

File: packages/marshal/src/encodeToCapData.js [hunk 1]
```diff
@@ -432,16 +433,9 @@ export const makeDecodeFromCapData = ({
       return result;
     } else {
       assert(typeof jsonEncoded === 'object' && jsonEncoded !== null);
```
Comment (gibson042) [inline packages/marshal/src/encodeToCapData.js:435]: Should we keep the no-symbol-keys assertion?
```suggestion
      assert(typeof jsonEncoded === 'object' && jsonEncoded !== null);
      for (const name of ownKeys(jsonEncoded)) {
        typeof name === 'string' ||
          assert.fail(X`Property ${name} of ${jsonEncoded} must be a string`);
      }
```
or in a single pass:
```suggestion
      assert(typeof jsonEncoded === 'object' && jsonEncoded !== null);
      const decodedEntries = ownKeys(jsonEncoded).map(name =>
        typeof name === 'string' ||
          assert.fail(X`Property ${name} of ${jsonEncoded} must be a string`);
        return [name, decodeFromCapData(jsonEncoded[name])];
      );
      return fromEntries(decodedEntries);
```

**Review by gibson042 — APPROVED** [review]
Suggested a few tweaks/renames and a Smallcaps test, but no need for re-review from me.

File: packages/marshal/src/encodeToCapData.js [hunk 2]
```diff
@@ -432,16 +433,9 @@ export const makeDecodeFromCapData = ({
       return result;
```
[the remark this answers is not in the document]
Comment (erights) [reply packages/marshal/src/encodeToCapData.js:433]: Good suggestions. Done.

File: packages/marshal/src/encodeToCapData.js [hunk 3]
```diff
@@ -432,16 +433,9 @@ export const makeDecodeFromCapData = ({
...
-          name,
-          'string',
-          X`Property ${name} of ${jsonEncoded} must be a string`,
-        );
-        result[name] = decodeFromCapData(jsonEncoded[name]);
-      }
-      return result;
+      const toDecEntry = ([name, subEnc]) => [name, decodeFromCapData(subEnc)];
+      const decEntries = entries(jsonEncoded).map(toDecEntry);
+      return fromEntries(decEntries);
```
[the remark this answers is not in the document]
Comment (erights) [reply packages/marshal/src/encodeToCapData.js:438]: Done

File: packages/marshal/src/encodeToSmallcaps.js [hunk 4]
```diff
@@ -471,17 +471,13 @@ export const makeDecodeFromSmallcaps = ({
           return result;
         }
 
-        const result = {};
-        for (const encodedName of ownKeys(encoding)) {
-          // TypeScript confused about `||` control flow so use `if` instead
-          // https://github.com/microsoft/TypeScript/issues/50739
-          if (typeof encodedName !== 'string') {
+        const toDecEntry = ([encodedName, subEnc]) => {
```
[the remark this answers is not in the document]
Comment (erights) [reply packages/marshal/src/encodeToSmallcaps.js:474]: Done

[the remark this answers is not in the document]
Comment (erights) [reply packages/marshal/src/encodeToSmallcaps.js:474]: Done

File: packages/marshal/src/encodeToSmallcaps.js [hunk 5]
```diff
@@ -491,9 +487,10 @@ export const makeDecodeFromSmallcaps = ({
             assert.fail(
               X`Decoded property name ${name} from ${encoding} must be a string`,
             );
-          result[name] = decodeFromSmallcaps(encoding[encodedName]);
-        }
-        return result;
+          return [name, decodeFromSmallcaps(subEnc)];
+        };
+        const decEntries = entries(encoding).map(toDecEntry);
+        return fromEntries(decEntries);
```
[the remark this answers is not in the document]
Comment (erights) [reply packages/marshal/src/encodeToSmallcaps.js:493]: Done

File: packages/marshal/src/encodeToCapData.js (hunk 1 above)
Comment (erights) [reply packages/marshal/src/encodeToCapData.js:435]: Done

File: packages/marshal/test/test-marshal-capdata.js [hunk 6]
```diff
@@ -371,3 +380,29 @@ test('records', t => {
...
+  );
+
+  const wrongProto = harden({ ['__proto__']: exampleAlice });
+  const wrongProtoEnc = ser(wrongProto);
+  t.deepEqual(wrongProtoEnc, {
+    body: '{"__proto__":{"@qclass":"slot","iface":"Alleged: Alice","index":0}}',
+    slots: ['slot'],
+  });
+  // Fails before https://github.com/endojs/endo/issues/1303 fix
+  t.deepEqual(unser(wrongProtoEnc), wrongProto);
```
[the remark this answers is not in the document]
Comment (erights) [reply packages/marshal/test/test-marshal-capdata.js]: Done

**Pull request merged**